[E2E] Tests for 'key'

WestleyK opened the pull request:
## Description of the Pull Request (PR):

This PR adds E2E test for the following key command:

1. `list`
2. `export`
3. `import`
4. `remove`
5. `pull`
6. `newpair`
7. `search`
8. ~`push`~ Not until CCI/Travis has an access token...

*Which is _almost_ all the `key` commands :tada:*

## This fixes or addresses the following GitHub issues:

- Fixes #3533 
- Fixes #2839

<br>

This PR must be merged after https://github.com/sylabs/singularity/pull/3575.

<br>

Commit:
Start for the e2e key tests

Commit:
Updated tests

Commit:
Updated tests

Commit:
Added pull/remove public key

Commit:
Start to add import tests

Commit:
Added import tests

Commit:
Added test for corrupted key

Commit:
Cleaned code

WestleyK commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,89 @@
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package e2e
```
This file is added here to provide some `singularity key [...]` commands.

mem replied:
Does it belong in e2e/internal/e2e, though? It's very specific to key testing.

Commit:
Added comment & updated key

Commit:
Fixed comments & Updated corruptKey func

Commit:
Added the start of testing private keys, Holy SH!!

jchavezlavalle commented on e2e/suite.go:
```diff
@@ -16,14 +16,15 @@ import (
 	"syscall"
 	"testing"
 
-	"github.com/sylabs/singularity/e2e/actions"
```
need to uncomment some lines around here and just add `"github.com/sylabs/singularity/e2e/key"` ?

jchavezlavalle commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// RecuverSecretKeyring ...
+func RecuverSecretKeyring(t *testing.T) {
```
What does this function do? @WestleyK

WestleyK replied:
This function `RecoverSecretKeyring` will recover your old keyring. When the tests run you keyring is backup to a tmp file, after the tests run your keyring is recovered.

jchavezlavalle replied:
okay good, there's a typo in this function name

jchavezlavalle commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,206 @@
+
+	s := getE(0, kpath, "", "e2etests")
+
+	t.Log("EEEEEEEEEEEEEEEEEE: ", s)
```
remove this log

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,131 @@
+	output := strings.Join(lines, "\n")
+	err = ioutil.WriteFile(kpath, []byte(output), 0644)
+	if err != nil {
+		t.Fatalf("Unabel to write to file: %v", err)
```
typo on word "Unable"

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,131 @@
+					if err == nil {
+						t.Log(string(out))
+						//t.Fatalf("Unexpected success when running: %s", cmd)
+					}
```
remove commented code

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,131 @@
+	t.Run("pull_default_key", test.WithoutPrivilege(func(t *testing.T) { e2e.PullDefaultPublicKey(t) }))
+
+	// Run the tests
+	t.Run("pubic_key", testPrivateKey)
```
typo on word public

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,131 @@
+	t.Run("pubic_key", testPrivateKey)
+
+	// Recuver the secret keyring
+	t.Run("recuvering_secret_keyring", test.WithoutPrivilege(func(t *testing.T) { e2e.RecuverSecretKeyring(t) }))
```
maybe typo as well on "Recovering" ?

jchavezlavalle commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+	output := strings.Join(lines, "\n")
+	err = ioutil.WriteFile(kpath, []byte(output), 0644)
+	if err != nil {
+		t.Fatalf("Unabel to write to file: %v", err)
```
typo on Unable

jchavezlavalle commented on e2e/suite.go:
```diff
@@ -89,19 +90,21 @@ func Run(t *testing.T) {
+	t.Run("KEYCMD", key.RunE2ETests)
 
-	t.Run("ACTIONS", actions.RunE2ETests)
+	//	t.Run("BUILD", imgbuild.RunE2ETests)
```
need as well to uncomment all these lines from previous tests

jchavezlavalle reviewed (changes requested):
left some comments, thanks @WestleyK

jchavezlavalle commented on e2e/key/key.go:
```diff
@@ -0,0 +1,102 @@
+		t.Fatal(err.Error())
+	}
+
+	t.Run("GenralKeyCmdTest", testGenralKeyCmd)
```
There's a typo in `GeneralKeyCmdTest` and `testGeneralKeyCmd`

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,131 @@
+var keyPath string
+var defaultKeyFile string
+
+//  coruptKey will take a ASCII key (kpath) and change some chars in it (corrupt it).
```
typo: corruptKey

Commit:
Addresed some comments & improved private key testing

Commit:
Updated funcs names + updated godoc commets & cleaned extra code

Commit:
Removed extra var, for CCI

Commit:
Fixed types++

Commit:
Added panic & corrupted import tests

Commit:
Added back tests

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+func PullDefaultPublicKey(t *testing.T) {
+	LoadEnv(t, &testenv)
+
+	argv := []string{"key", "pull", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
```
this is working against ~/.singularity/sypgp.

I can see that you have this set up with backup / restore to make sure that this key is not already present in that directory.

I would prefer it if the code (`DirPath` in `pkg/sypgp/sypgp.go` and related functions) is modified so that it either uses an environment variable to obtain the name of the `.singularity` directory, or it uses an environment variable to obtain the name of the `.singularity/sypgp` directory. Then you set this environment variable to a temporary directory that you know for a fact that does NOT contain this key.

I prefer the first option even if it's much more work, but the second is acceptable for now.

jchavezlavalle commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+// RecoverSecretKeyring will recover your secret keyring, this gets ran after the private key test are complete.
```
gets **run**

jchavezlavalle commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+// RemoveSecretKeyring will delete your secret keyring :O
```
remove surprised emoji at the end of this comment lol

jchavezlavalle commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,208 @@
+	return execKey.CombinedOutput()
+}
+
+// ImportTestKey will take a private key file (kpath) and import it.
```
Update comment: ImportTestKey -> ImportPrivateKey

jchavezlavalle commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+	t.Run("pull_default_key", test.WithoutPrivilege(func(t *testing.T) { e2e.PullDefaultPublicKey(t) }))
+
+	// Run the tests
+	t.Run("pubic_key", testPublicKey)
```
typo on public_key

jchavezlavalle reviewed (changes requested):
looking good, added some comments around @WestleyK

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+func RemoveDefaultPublicKey(t *testing.T) {
+	LoadEnv(t, &testenv)
+
+	argv := []string{"key", "remove", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
```
even if the above is a nitpick, this is not.

How do you ensure that the key exists? This needs fixing.

WestleyK replied:
First, the command will not fail if theres no key. Second, if there is no key when this function is called, theres a another problem. So I dont think its necessary to check if the key exist first...

mem replied:
Uhm...

in order to help with automation, the command _should_ return with a non-zero exit code if the key to be removed does not exist. If that's currently not the case, an issue should be opened for that, and a test should be added, so that we the code changes, the test catches it.

As to the "other problem"... what might that be?

My original point is that in a test you should make sure that all the preconditions for the test are met. If I see this test without any other context, I would expect that there's something making sure this particular key exists. I don't see that here, which makes the test code harder to understand.

WestleyK replied:
>First, the command will not fail if theres no key.

Nevermined, it does fail if theres no key.

Any ideas on how to check if this key is in the keyring?

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+		t.Fatal(err.Error())
+	}
+
+	t.Run("backingup_secret_keyring", test.WithoutPrivilege(func(t *testing.T) { e2e.BackupSecretKeyring(t) }))
```
this is not a test, this should not be running with t.Run

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	}
+}
+
+// TestAll is trigered by ../key.go, that is trigered by suite.go in the e3e test directory
```
this is documenting how the function works (internal documentation) not what the function does (external documentation). You can move this inside the function.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	t.Run("importing_test_key", test.WithoutPrivilege(func(t *testing.T) {
+		b, err := e2e.ImportPrivateKey(t, "./key/testdata/e2e_test_key.asc")
+		if err != nil {
+			t.Log(string(b))
```
Please add something to the output what indicates what it is, for example, "output of running <whatever>:\n"

mem commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+		t.Fatal(err.Error())
+	}
+
+	keyPath = testenv.TestDir
```
please don't use globals here.

WestleyK replied:
? This is in its own package, is it imported not to use globals? :thinking:

mem commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	keyPath = testenv.TestDir
+	defaultKeyFile = filepath.Join(keyPath, "exported_key")
```
please don't use globals here.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+		}
+	}))
+
+	keyPath = testenv.TestDir
```
please don't use globals here.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	}))
+
+	keyPath = testenv.TestDir
+	defaultKeyFile = filepath.Join(keyPath, "exported_private_key")
```
please don't use globals here.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	t.Run("public_key", testPrivateKey)
+
+	// Recover the secret keyring
+	t.Run("recovering_secret_keyring", test.WithoutPrivilege(func(t *testing.T) { e2e.RecoverSecretKeyring(t) }))
```
this is also not a test, this is teardown.

If you need to set up and tear down for each test, you can wrap the second argument to t.Run with a function that takes a func (t *testing.T) function and returns a func (t *testing.T) function, just like test.WithoutPrivilege does.

For example:

```
func withEmptySecretKeyring(f func (t *testing.T)) {
    return func (t *testing.T) {
        backupSecretKeyring(t)
        defer restoreSecretKeyringt(t)
        f(t)
    }
}
```

`defer` is required becuase defer runs even if the function f panics.

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+`, armor, kpath, num, psk)
+}
+
+var backupSypgp = filepath.Join(HomeDir(), ".singularity/sypgp/secret-keyring-backup")
```
there's an exported function in the sypgp package that returns "~/.singularity/sypgp".

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+var backupSypgp = filepath.Join(HomeDir(), ".singularity/sypgp/secret-keyring-backup")
+
+// PullDefaultPublicKey will pull the public Sylabs Admin key
+func PullDefaultPublicKey(t *testing.T) {
```
this does not need to be exported.

WestleyK replied:
Umm, it used in `e2e/key/keypublic/keypublic.go` to pull the key.

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	argv := []string{exportScript.Name()}
+	execImport := exec.Command("/usr/bin/expect", argv...)
```
is it necessary to hardcode the path like this?

WestleyK replied:
Well, it seems to work without the full path :smile:

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+	argv := []string{exportScript.Name()}
+	execImport := exec.Command("/usr/bin/expect", argv...)
+
+	return execImport.CombinedOutput()
```
you should check the exit code of the command, no?

WestleyK replied:
This is checked where the function is called, in some test cases, this is supposed to fail.

mem commented on e2e/internal/e2e/keyexec.go:
```diff
@@ -0,0 +1,204 @@
+		t.Fatalf("%v", err)
+	}
+
+	execKey.Stdin = stdinRun
```
note that Cmd.Stdin is an io.Reader. A bytes.Buffer is also an io.Reader (also strings.Reader).

This means you can place `stdin` in a bytes.Buffer and you are set. Since `stdin` is a `string` and not a `[]byte`, you can use `strings.NewReader` to obtain an io.Reader.

mem commented on e2e/key/key.go:
```diff
@@ -0,0 +1,102 @@
+		succeed bool
+	}{
+		{
+			name:    "key_list",
```
you omit the underscore. When running with -v that will read more naturally.

WestleyK replied:
Okay, I'll remove the `_` :+1:

mem commented on e2e/key/key.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	for _, tt := range tests {
+		t.Run("key_run "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
```
no need for the "key_run" prefix, as it will automatically get prefixed with the subtest that is calling this.

mem commented on e2e/key/key.go:
```diff
@@ -0,0 +1,102 @@
+	for _, tt := range tests {
+		t.Run("key_run "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
+			cmd, out, err := e2e.RunKeyCmd(t, testenv.CmdPath, tt.args, tt.file, "")
+			if tt.succeed {
```
please see e2e/pull/pull.go around line ~ 302 (switch) for an example of how to handle this. You need to consider four cases: expecting failure + failed; expecting success + success; expecting failure + success; expecting success + failure and my recommendation is to use a switch to make it explicit that you are handling all four cases.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	RunDisabled bool   `default:"false"`
+}
+
+var testenv testingEnv
```
please no globals in tests.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+var defaultKeyFile string
+
+// corruptKey will take a ASCII key (kpath) and change some chars in it (corrupt it).
+func corruptKey(t *testing.T, kpath string) {
```
this is a good idea for a test, but my gut feeling is that this is better suited for a unit test.

Is there something that cannot be tested as a unit test?

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+		{
+			name:    "export_private",
+			args:    []string{"export", "--secret"},
+			stdin:   "0\n", // TODO: this will need to be '1' at some point in time -> issue #3199
```
can you be more explicit as to what's the meaning of that 0 and why it's changing to a 1?

WestleyK replied:
Take a look at this issue https://github.com/sylabs/singularity/issues/3199.

mem replied:
I know I can look at the issue. The issue is too far away when I'm looking at code. It's good to document that there's an open issue related to this, but it's not enough.

In this specific case, something like "the first entry in the selection list is item 0; see issue #3199"

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	for _, tt := range tests {
+		t.Run("key_run "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
```
same here with "key_run"

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,136 @@
+		t.Run("key_run "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
+			os.RemoveAll(filepath.Join(keyPath, defaultKeyFile))
+			out, err := e2e.ExportPrivateKey(t, tt.file)
+			if tt.succeed {
```
same here

mem commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+var defaultKeyFile string
+
+// corruptKey will take a ASCII key (kpath) and change some chars in it (corrupt it).
+func corruptKey(t *testing.T, kpath string) {
```
this looks like it's the same code as the other package

mem commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+		succeed bool
+	}{
+		{
+			name:    "export_public",
```
same comment with the underscores.

mem commented on e2e/key/keypublic/keypublic.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	for _, tt := range tests {
+		t.Run("key_run "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
```
same with the "key_run" label

Commit:
Updated for comments

Commit:
Added test case

Commit:
[TEST] testing out the go-expect package ??? WIP

Commit:
Adding vendor

Commit:
Removed vednor

Commit:
Added Netflix vednor

Commit:
Trying to add vendor

Commit:
Fixed some test cases

Commit:
Added back case err check

Commit:
Chenged struct, bools first, then strings

Commit:
Ops, strings first, then bools

Commit:
Removed unused code, -> CCI

Commit:
Fixed ImportPrivateKey()

Commit:
Moved keyexec.go to its own package

Commit:
Cleaned up code

Commit:
Almost fully working tests!

Commit:
Cleaned up functions

Commit:
Dont print stdin to stdout, YES

WestleyK commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// CorruptKey will take a ASCII key (kpath) and change some chars in it (corrupt it).
+func CorruptKey(t *testing.T, kpath string) {
```
Would it be better to use `sed` and call a `exec.Command()`?

WestleyK replied:
Or even better is to put a already corrupted key in the `e2e/key/testdata`?... <- I vote for this.

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Removed commented code

Commit:
Quick save 1

Commit:
Updated docs

Commit:
Merge branch 'sygpg-dir' into e2e-key

Commit:
Resolving conflict

Commit:
Use SINGULARITY_SYPGPDIR

Commit:
Removed unused functions

Commit:
Trying to update go.sum

mem commented on e2e/internal/e2e/env.go:
```diff
@@ -15,3 +16,8 @@ func LoadEnv(t *testing.T, env interface{}) {
 }
+
+// HomeDir will return the users home directory.
+func HomeDir() string {
```
by default, please do not export functions. It increases the size of the API and makes things more difficult to use.

It's perfectly fine to export functions if you need to use the code from multiple packages, I just can't find where this is being used, so I can't really say if this needs to be exported or not.

WestleyK replied:
Actually, its not being used anymore, I'll just delete it.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,216 @@
+
+// PullDefaultPublicKey will pull the public Sylabs Admin key
+func PullDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
```
this function is not using values from testenv, so you don't need this.

WestleyK replied:
Wait, it does actually use it. I dont know why CCI/Travis passed, it failed locally for me.

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Adding test for key search/push

Commit:
Adding newpair testing

Commit:
Fixed the tests for newpair

Commit:
Cleaning up extra code

Commit:
Added back tests

Commit:
Added back tests p.2

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+func PullDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "pull", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
```
since the fingerprint is going to be used multiple times across the test (or across multiple tests?) can you please put this is a package-level constant?

Commit:
Removed extra function & removed unnessary e2e.LoadEnv...

ArangoGutierrez committed:
Merge branch 'master' into e2e-key

Commit:
Added back e2e.LoadEnv()

Commit:
Merge branch 'e2e-key' of github.com:WestleyK/singularity into e2e-key

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,202 @@
+				if err != nil {
+					t.Log("Command that failed: ", c)
+					t.Log(string(b))
+					t.Fatalf("unepexted failure: %v", err)
```
typo on word unexpected

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,202 @@
+					t.Log(string(b))
+					t.Fatalf("unepexted failure: %v", err)
+				}
+				//t.Run("testing_new_key_from_newpair", keyexec.QuickTestKey)
```
remove line 71 commented code

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,202 @@
+			} else {
+				if err == nil {
+					t.Log(string(b))
+					t.Fatalf("unexpected succees running: %v", err)
```
typo on line 76 word success

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+	// Send the passcode to singularity
+	c.Send("e2etests\n")
+	c.Send("e2etests\n")
+	c.Send("e2etests\n")
```
why 3 times?

WestleyK replied:
Because when you import a private key, it will ask you for the old password, then a new password, and then to retype you password. For this test, its all using the same password.

jchavezlavalle replied:
cool, could you add a comment to that? so that when somebody needs to maintain it in the future it could be clear

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+	cmd.Stdin = c.Tty()
+
+	cmd.Stderr = outErr
+	cmd.Stdout = outErr
```
why are these 2 sent as error and output with the same return value?

WestleyK replied:
This is to capture the stdout, and stderr, and returns it. Useful for debugging or when a test fails.

jchavezlavalle replied:
why not using the combined output instead?

WestleyK replied:
Because `cmd.CombinedOutput` will wait for the command to finish running, and we want to use `c.Send()` to send input for the command thats running, (we cant do that if the command is already done running). So here we use `cmd.Start()` to start the command, then send the input, and then wait for the command to finish running by `cmd.Wait()`. *hope that makes some sense...*

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+
+// QuickTestKey will export a private, and public key (0), and then import them. This is used after
+// generating a newpair.
+func QuickTestKey(t *testing.T) {
```
Better to call it `QuickTestExportImportKey` to know exactly what this test is doing beforehand

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+			succeed: true,
+		},
+		{
+			name:    "quick test public  armor",
```
additional space on name (between public and armor)

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+				if err != nil {
+					t.Log("Command that failed: ", c)
+					t.Log(string(b))
+					t.Fatalf("unepexted failure: %v", err)
```
typo on line 290

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+			} else {
+				if err == nil {
+					t.Log(string(b))
+					t.Fatalf("unexpected succees running: %v", c)
```
typo on line 307

jchavezlavalle commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,361 @@
+	}
+}
+
+func KeyNewPair(t *testing.T, user, email, note, psk1, psk2 string, push bool) (string, []byte, error) {
```
add comments on all functions before the definition, so that it can be clear what they're doing :+1:

jchavezlavalle commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,202 @@
+
+var testenv testingEnv
+var keyPath string
+var defaultKeyFile string
```
remove globals `keyPath` and `defaultKeyFile` from global scope and define them where needed

Commit:
Fixed for reviewers

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Fixed misspell

Commit:
Fixed a couple things here and there...

Commit:
Remove the 'push' tests

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Added more control test keys to import.

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Fixed coruppted tests, moved to key.go

Commit:
Updated tests with control keys

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+var testenv testingEnv
+
+// PullDefaultPublicKey will pull the public Sylabs Admin key
+func PullDefaultPublicKey(t *testing.T) {
```
looks to me that this is not really a test, but a helper function, maybe don't export it?

WestleyK replied:
Its used in `e2e/key/keypublic/keypublic.go:172`.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// RemoveDefaultPublicKey will pull the public Sylabs Admin key
+func RemoveDefaultPublicKey(t *testing.T) {
```
same here about not exporting

WestleyK replied:
It is used in `e2e/key/keypublic/keypublic.go:144`. Should I just put the whole `RemoveDefaultPublicKey` function in keypublic.go?

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+func RemovePublicKeyring(t *testing.T) {
+	err := os.Remove(sypgp.PublicPath())
```
for this to succeed, sypgp.PublicPath() ist be empty. Maybe document this as part of the function's documentation?

WestleyK replied:
`sypgp.PublicPath()` should return a file, not a directory.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+func RemovePublicKeyring(t *testing.T) {
```
don't export

WestleyK replied:
Okay, this function is only used in this file :+1:

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// RemoveSecretKeyring will delete your secret keyring.
+func RemoveSecretKeyring(t *testing.T) {
```
don't export

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+
+// RemoveSecretKeyring will delete your secret keyring.
+func RemoveSecretKeyring(t *testing.T) {
+	err := os.Remove(sypgp.SecretPath())
```
same comment regarding sypgp.SecretPath() having to be empty.

WestleyK replied:
Also should return a file.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+func RemoveKeyring(t *testing.T) {
```
don't export

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// ImportKey will import a key from kpath.
+func ImportKey(t *testing.T, kpath string) (string, []byte, error) {
```
don't export

I'll stop adding this comment. Most (all?) the functions don't need to be exported.

WestleyK replied:
But these functions are used in different files, like `e2e/key/keypublic/keypublic.go`, Should I put these function(s) in that file? _that would make it very big..._

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+
+	b, err := execKey.CombinedOutput()
+
+	return cm, b, err
```
shouldn't you call t.Error if err != nil?

WestleyK replied:
I dont think so... Because in some tests, we are expecting it to fail.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+	exportCmd := []string{"key", "import", kpath}
+
+	cmd := exec.Command(testenv.CmdPath, exportCmd...)
+	cmd.Stdin = c.Tty()
```
why?

WestleyK replied:
Why to what? Connecting `cmd.Stdin` to `c.Tty()` ?

WestleyK replied:
This makes it possible to use `c.Send()` to provide a standard input to the command, to my knowledge...

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+
+	// Send the passcode to singularity. The first one is the old
+	// password, the next two are the new passowrd.
+	c.Send("e2etests\n")
```
the key you are importing will always have this passphrase? Shouldn't this be a parameter to the function?

WestleyK replied:
For simplicity its I hard coded it _for now_. Right now its not testing passwords/wrong passwords. Should I change this now?

mem replied:
you can place this value (`e2etests`, without the newline) in a variable so that at least there's a hint that this will change / needs to change.

WestleyK replied:
Yes, I should of done that before, but it does need the newline. I'll add a comment about that.

mem replied:
you can add the new line as part of sending the password using expect.

What I mean is that the password does not contain the newline, so to make it clear that the variable contains the password (and not something else), it should not contain the newline.

WestleyK replied:
Got it! okay :+1:

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// RunKeyCmd will run a 'singularty key' command, with any args that are set in commands.
+func RunKeyCmd(t *testing.T, cmdPath string, commands []string, file, stdin string) (string, []byte, error) {
```
why is the "file" argument special? It's used only once. The user of this function could pass it as part of the arguments (commands)

WestleyK replied:
Its slightly easier to read in the table tests, imo. But it could be part of the `commands []string`.

WestleyK replied:
Fixed now! :+1:

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -0,0 +1,347 @@
+			if tt.private {
+				c, b, err = ExportPrivateKey(t, filepath.Join(tmpTestDir, "export_key.asc"), "0\n", tt.armor)
+			} else {
+				c, b, err = RunKeyCmd(t, testenv.CmdPath, []string{"export"}, filepath.Join(tmpTestDir, "export_key.asc"), "0\n")
```
what's the 0 entry?

the bit that is confusing is what does the keyring contain when this is running and why?

WestleyK replied:
Key 0 is the newly created key by `newpair`. The keyring is cleared before generating a newpair. This function just tests that key from `newpair` can do what its supposed to.

mem replied:
please add a comment indicating that the keyring is empty before running this and that the newly added key gets to be entry 0 in all cases. Also, I believe there's an issue for the keys being zero-based, no?

WestleyK replied:
Okay, I changed this test to export all the keys first, then try to import them, one by one. This way the keyring _could_ still have another entity in it.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,166 @@
+			user:    "genbye2etests\n",
+			email:   "westley@sylabs.io\n",
+			note:    "test key generated by e2e tests\n",
+			psk1:    "e2etests\n",
```
psk2 is not set, is that intentional?

WestleyK replied:
Yes, if `psk2` is not set, `keyexec.KeyNewPair()` will use the same password as `psk1`.

mem replied:
uhg.

:-)

WestleyK replied:
Okay, I'll remove `psk2`, its not necessary right now.

mem commented on e2e/key/keyprivate/keyprivate.go:
```diff
@@ -0,0 +1,166 @@
+					t.Log(string(b))
+					t.Fatalf("unexpected failure: %v", err)
+				}
+				keyexec.QuickTestExportImportKey(t)
```
I'm going to guess that QuickTestExportImportKey _depends_ on being called exactly here.

What I mean is that since this function receives only t as parameter, it's tightly coupled with the key ring that is being created or modified here.

WestleyK replied:
I also included another parameter for this function, that will tell `QuickTestExportImportKey` what key to test.

Commit:
Added e2e test key fingerprint as a exported const

Commit:
Unexported a function that did not need to be exported

Commit:
Added "newline" to "command that failed"
To make it easier to read and debug

Commit:
Merge remote-tracking branch 'upstream/master' into e2e-key

Commit:
Added file to command []string, use const keyPassword

Commit:
Removed used test code + added more comments ...
Added number-for-entity to test when testing newpair

Commit:
When running the quickExportImport tests, this will now export
all of the keys, then import the keys one by one. This way the
keyring file can have more then one entity, but after the test
is ran, there will be no keys in the tmp-keyring file. And when
calling the QuickExportImport test, there is a argument for
witch key you want to test. + Uncaped some error messages

Commit:
Uncaped _all_ the error messages + use common const password
for all functions using the password.

Commit:
Fixed struct size for CCI :roll_eyes:

Commit:
Dont run "remove private key before importing" with t.Run
Now it does not "hang" for a couple mili-seconds, like before +
improved error output logging.

mem commented on e2e/internal/keyexec/keyexec.go:
```diff
@@ -172,118 +177,127 @@ func ExportPrivateKey(t *testing.T, kpath, num string, armor bool) (string, []by
-// generating a newpair.
-func QuickTestExportImportKey(t *testing.T) {
+// generating a newpair. keyNum is the key number to test. Its a string, so the number must be used in "", and
+// end with a '\n', eg. "1\n" will test key 1.
```
it would be better to make keyNum an int, and have this function convert it to a string and add the necessary newline.

Commit:
Changed key number from string, to int, as @mem suggested :+1:

Commit:
Consolidated some basic importing tests + remove the tmp-sypgp_dir before re-importing
I have a feelling I will need to re-write ~some~ **most** of the tests, but I'll push my changes so they dont get lost.
